src/test/shared/clients/defaultS3Client.test.ts

```diff
@@ -108,6 +165,59 @@ describe('DefaultS3Client', () => {
         })
     })
 
+    describe('deleteBucket', () => {
+        const {
+            firstPageRequest: firstList,
+            secondPageRequest: secondList,
+            firstPageResponse: firstListResponse,
+            secondPageResponse: secondListResponse,
+        } = new ListObjectVersionsFixtures()
+        const anyListResponse = secondListResponse
+        const { firstRequest: firstDelete, secondRequest: secondDelete } = new DeleteObjectsFixtures()
+
+        it('empties a bucket and deletes it', async () => {
+            when(mockS3.listObjectVersions(deepEqual(firstList))).thenReturn(success(firstListResponse))
+            when(mockS3.deleteObjects(deepEqual(firstDelete))).thenReturn(success({}))
+
+            when(mockS3.listObjectVersions(deepEqual(secondList))).thenReturn(success(secondListResponse))
+            when(mockS3.deleteObjects(deepEqual(secondDelete))).thenReturn(success({}))
+
+            when(mockS3.deleteBucket(deepEqual({ Bucket: bucketName }))).thenReturn(success({}))
+
+            await createClient().deleteBucket({ bucketName })
+
+            verify(mockS3.listObjectVersions(anything())).twice()
+            verify(mockS3.deleteObjects(anything())).twice()
+            verify(mockS3.deleteBucket(anything())).once()
```

Comment:
What is this testing, other than that the implementation didn't change?

Reply:
It's verifying that the mocks are actually called, which is important in this case. Even if you set up the mocks, the tests could pass without invoking them. There's no "verify all my stubs were invoked" lifecycle check.

e.g. if someone removed the `deleteBucket` call from the function, this test would still pass without the verification step.

Reply:
Quite necessary when you can't verify the output of the function for correctness of the series of service calls.

Reply:
I'm referring to the entire test (and the other tests). The `when()` calls set up a thing unrelated to the implementation. The thing happens to match the implementation. Then we `verify()` that the thing does something. This is a lot of test code that ultimately tests very little.

Reply:
Not sure what you mean. It tests everything apart from the actual S3 service which would require an end-to-end test. These are unit tests for this class's logic, not tests for the real S3. They test the algorithm this class uses to delete a bucket. They test that this class doesn't delete the wrong files or the wrong bucket which would be really bad. How else would you test this without invoking the real service?

Reply:
We have fakes; I would consider adding logic to the fakes to mimic S3 with an in-memory collection. Though that might be worse.  

>  They test that this class doesn't delete the wrong files or the wrong bucket which would be really bad.

I see tests for when  `listObjectVersions` etc fail. But the tests depend on the impl staying the same and calling the same functions internally. 

I guess it's better than nothing, and is worth the extra assertions even if they are thin.

Reply:
> We have fakes; I would consider adding logic to the fakes to mimic S3 with an in-memory collection. Though that might be worse.

Yeah that would work fine and would be especially useful if the underlying fake were being used in multiple places so that the users wouldn't have to rewrite the mocks. 

However, in this case it seemed appropriate to use targeted mocked calls to remove the need for the somewhat complex state management to maintain a proper fake. The goal here in the wrapper is also to simplify the interactions with the service to confine these types of complicated mocks to this class's unit tests. Callers should just need to interact with the wrapper. If the wrapper isn't providing a clearer and more testable interface than the raw service, than the abstraction isn't working well.

It'd be great to use some existing S3 in memory fakes, but I'm not aware of any comprehensive and reliable ones. DynamoDB has dynamodb local which is great for this sort of thing.

> I see tests for when listObjectVersions etc fail. But the tests depend on the impl staying the same and calling the same functions internally.

Yeah the downside of mocking being that it couples you to those calls. But the upside is that you can verify side effects that don't produce any output without defining, maintaining, and verifying the underlying state.